gc/src/gc.rs

```diff
@@ -270,58 +270,56 @@ fn collect_garbage(st: &mut GcState) {
         // Walk the tree, tracing and marking the nodes
         let mut mark_head = head.get();
         while let Some(node) = mark_head {
-            if node.as_ref().header.roots() > 0 {
-                node.as_ref().trace_inner();
+            if unsafe { node.as_ref().header.roots() } > 0 {
+                unsafe { node.as_ref().trace_inner() };
             }
 
-            mark_head = node.as_ref().header.next.get();
+            mark_head = unsafe { node.as_ref().header.next.get() };
         }
 
         // Collect a vector of all of the nodes which were not marked,
         // and unmark the ones which were.
         let mut unmarked = Vec::new();
         let mut unmark_head = head;
         while let Some(node) = unmark_head.get() {
-            if node.as_ref().header.is_marked() {
-                node.as_ref().header.unmark();
+            if unsafe { node.as_ref().header.is_marked() } {
+                unsafe { node.as_ref().header.unmark() };
             } else {
                 unmarked.push(Unmarked {
                     incoming: unmark_head,
                     this: node,
                 });
             }
-            unmark_head = &node.as_ref().header.next;
+            unmark_head = unsafe { &node.as_ref().header.next };
         }
         unmarked
     }
 
     unsafe fn sweep(finalized: Vec<Unmarked<'_>>, bytes_allocated: &mut usize) {
         let _guard = DropGuard::new();
         for node in finalized.into_iter().rev() {
-            if node.this.as_ref().header.is_marked() {
+            if unsafe { node.this.as_ref().header.is_marked() } {
                 continue;
             }
             let incoming = node.incoming;
-            let node = Box::from_raw(node.this.as_ptr());
+            let node = unsafe { Box::from_raw(node.this.as_ptr()) };
             *bytes_allocated -= mem::size_of_val::<GcBox<_>>(&*node);
             incoming.set(node.header.next.take());
         }
     }
 
     st.stats.collections_performed += 1;
 
-    unsafe {
-        let head = Cell::from_mut(&mut st.boxes_start);
-        let unmarked = mark(head);
-        if unmarked.is_empty() {
-            return;
-        }
-        for node in &unmarked {
-            Trace::finalize_glue(&node.this.as_ref().data);
-        }
-        mark(head);
-        sweep(unmarked, &mut st.stats.bytes_allocated);
+    let head = Cell::from_mut(&mut st.boxes_start);
```

Comment:
issue: ditto here, I think having this whole block be `unsafe` was more readable